File: apps/desktop/src-tauri/src/services/git.rs
use std::path::Path;
use std::process::Command;

use crate::models::workspace::{WorkspaceCommit, WorkspaceGitInfo, WorkspaceGitStatus};

pub fn inspect_workspace(path: &str) -> WorkspaceGitInfo {
    let root = Path::new(path);
    if !root.is_dir() {
        return WorkspaceGitInfo {
            path_exists: false,
            is_repository: false,
            status: None,
            commits: Vec::new(),
            error: Some("项目目录不存在或不可访问".to_owned()),
        };
    }

    let status = Command::new("git")
        .args(["status", "--short", "--branch"])
        .current_dir(root)
        .output();
    let Ok(status_output) = status else {
        return unavailable("无法执行 git，请确认 Git 已安装", true);
    };
    if !status_output.status.success() {
        return WorkspaceGitInfo {
            path_exists: true,
            is_repository: false,
            status: None,
            commits: Vec::new(),
            error: None,
        };
    }

    let status = parse_status(&String::from_utf8_lossy(&status_output.stdout));
    let log_output = Command::new("git")
        .args([
            "log",
            "-10",
            "--date=iso-strict",
            "--format=%H%x1f%h%x1f%an%x1f%ad%x1f%s",
        ])
        .current_dir(root)
        .output();
    let commits = log_output
        .ok()
        .filter(|output| output.status.success())
        .map(|output| parse_commits(&String::from_utf8_lossy(&output.stdout)))
        .unwrap_or_default();

    WorkspaceGitInfo {
        path_exists: true,
        is_repository: true,
        status: Some(status),
        commits,
        error: None,
    }
}

fn unavailable(message: &str, path_exists: bool) -> WorkspaceGitInfo {
    WorkspaceGitInfo {
        path_exists,
        is_repository: false,
        status: None,
        commits: Vec::new(),
        error: Some(message.to_owned()),
    }
}

fn parse_status(output: &str) -> WorkspaceGitStatus {
    let mut branch = None;
    let mut ahead = 0;
    let mut behind = 0;
    let mut staged = 0;
    let mut modified = 0;
    let mut untracked = 0;
    let mut conflicted = 0;

    for (index, line) in output.lines().enumerate() {
        if index == 0 && line.starts_with("## ") {
            let tracking = line.trim_start_matches("## ");
            let mut parts = tracking.split("...");
            branch = Some(
                parts
                    .next()
                    .unwrap_or(tracking)
                    .split(" [ahead")
                    .next()
                    .unwrap_or(tracking)
                    .to_owned(),
            );
            if let Some(details) = tracking
                .split_once(" [")
                .map(|(_, value)| value.trim_end_matches(']'))
            {
                for item in details.split(", ") {
                    let mut values = item.split_whitespace();
                    let kind = values.next().unwrap_or_default();
                    let count = values
                        .next()
                        .and_then(|value| value.parse::<u32>().ok())
                        .unwrap_or(0);
                    if kind == "ahead" {
                        ahead = count;
                    }
                    if kind == "behind" {
                        behind = count;
                    }
                }
            }
            continue;
        }
        let bytes = line.as_bytes();
        if bytes.len() < 2 {
            continue;
        }
        let index_status = bytes[0] as char;
        let worktree_status = bytes[1] as char;
        if index_status == '?' && worktree_status == '?' {
            untracked += 1;
            continue;
        }
        if index_status == 'U'
            || worktree_status == 'U'
            || (index_status == 'A' && worktree_status == 'A')
        {
            conflicted += 1;
            continue;
        }
        if index_status != ' ' {
            staged += 1;
        }
        if worktree_status != ' ' {
            modified += 1;
        }
    }

    WorkspaceGitStatus {
        is_repository: true,
        branch,
        ahead,
        behind,
        staged,
        modified,
        untracked,
        conflicted,
        clean: staged == 0 && modified == 0 && untracked == 0 && conflicted == 0,
    }
}

fn parse_commits(output: &str) -> Vec<WorkspaceCommit> {
    output
        .lines()
        .filter_map(|line| {
            let fields: Vec<&str> = line.split('\u{1f}').collect();
            if fields.len() < 5 {
                return None;
            }
            Some(WorkspaceCommit {
                hash: fields[0].to_owned(),
                short_hash: fields[1].to_owned(),
                author: fields[2].to_owned(),
                date: fields[3].to_owned(),
                subject: fields[4..].join("\u{1f}"),
            })
        })
        .collect()
}

File: apps/server/src/platform/git.ts
import { execFile } from "node:child_process";
import { readFile, stat } from "node:fs/promises";
import path from "node:path";
import { promisify } from "node:util";
import type {
  WorkspaceGitCommitResult,
  WorkspaceGitDiff,
  WorkspaceGitFile,
  WorkspaceGitSummary,
} from "@chatdesk/shared";

type GitStatusFile = {
  status: WorkspaceGitFile["status"];
  previousPath?: string;
};

type GitStatusValues = {
  isRepository: true;
  branch: string | null;
  ahead: number;
  behind: number;
  staged: number;
  modified: number;
  untracked: number;
  conflicted: number;
  clean: boolean;
};

export type GitStatusSnapshot = GitStatusValues & {
  files: Map<string, GitStatusFile>;
};

const execFileAsync = promisify(execFile);
export const MAX_GIT_FILES = 200;
export const MAX_GIT_DIFF_BYTES = 128 * 1024;
const MAX_GIT_RESPONSE_BYTES = 1024 * 1024;

async function runGit(root: string, args: string[], maxBuffer = MAX_GIT_RESPONSE_BYTES) {
  return execFileAsync("git", args, {
    cwd: root,
    timeout: 8_000,
    maxBuffer,
  });
}

async function readGitSnapshot(root: string, filePath: string) {
  try {
    const result = await runGit(root, ["show", `HEAD:${filePath}`], MAX_GIT_DIFF_BYTES + 1);
    return result.stdout;
  } catch {
    return "";
  }
}

async function readWorkingSnapshot(root: string, filePath: string) {
  try {
    const metadata = await stat(path.join(root, filePath));
    if (metadata.size > MAX_GIT_DIFF_BYTES) return null;
    return await readFile(path.join(root, filePath), "utf8");
  } catch {
    return "";
  }
}

function safeRelativePath(root: string, relativePath: string) {
  const candidate = path.resolve(root, relativePath.trim());
  const relative = path.relative(root, candidate);
  if (!relative || relative.startsWith("..") || path.isAbsolute(relative)) {
    throw new Error("路径必须位于当前 workspace 内");
  }
  return relative;
}

function countLines(value: string) {
  if (!value) return 0;
  return value.split(/\r?\n/).length - (value.endsWith("\n") ? 1 : 0);
}

function statusForPath(status: string): WorkspaceGitFile["status"] {
  if (status.includes("U")) return "conflicted";
  if (status.includes("?") || status === "??") return "untracked";
  if (status.includes("R")) return "renamed";
  if (status.includes("A")) return "added";
  if (status.includes("D")) return "deleted";
  return "modified";
}

function parseBranchHeader(value: string) {
  const tracking = value.slice(3);
  const noCommits = tracking.match(/^No commits yet on (.+)$/);
  const branch = noCommits?.[1] ?? tracking.split("...")[0].split(" [ahead")[0] ?? null;
  let ahead = 0;
  let behind = 0;
  const details = tracking.match(/\[(.*?)\]/)?.[1] ?? "";
  for (const item of details.split(", ")) {
    const [kind, count] = item.split(/\s+/);
    if (kind === "ahead") ahead = Number(count) || 0;
    if (kind === "behind") behind = Number(count) || 0;
  }
  return { branch, ahead, behind };
}

function parseNumstat(output: string) {
  return output
    .split("\0")
    .filter(Boolean)
    .flatMap((record) => {
      const fields = record.split("\t");
      if (fields.length < 3) return [];
      const additions = fields[0] === "-" ? null : Number(fields[0]);
      const deletions = fields[1] === "-" ? null : Number(fields[1]);
      return [
        {
          path: fields[2],
          additions,
          deletions,
          binary: additions === null,
        },
      ];
    });
}

export async function readGitStatus(root: string): Promise<GitStatusSnapshot> {
  const result = await runGit(root, [
    "status",
    "--porcelain=v1",
    "--branch",
    "--untracked-files=all",
    "-z",
  ]);
  const tokens = result.stdout.split("\0").filter(Boolean);
  let branch: string | null = null;
  let ahead = 0;
  let behind = 0;
  let staged = 0;
  let modified = 0;
  let untracked = 0;
  let conflicted = 0;
  const files = new Map<string, GitStatusFile>();

  for (let index = 0; index < tokens.length; index += 1) {
    const token = tokens[index];
    if (token.startsWith("## ")) {
      ({ branch, ahead, behind } = parseBranchHeader(token));
      continue;
    }
    if (token.length < 4) continue;
    const status = token.slice(0, 2);
    const pathValue = token.slice(3);
    const fileStatus = statusForPath(status);
    let previousPath: string | undefined;
    if (status.includes("R") || status.includes("C")) {
      previousPath = tokens[index + 1];
      index += 1;
    }
    files.set(pathValue, { status: fileStatus, previousPath });
    if (status === "??") {
      untracked += 1;
      continue;
    }
    if (status[0] === "U" || status[1] === "U" || (status[0] === "A" && status[1] === "A")) {
      conflicted += 1;
      continue;
    }
    if (status[0] !== " ") staged += 1;
    if (status[1] !== " ") modified += 1;
  }

  return {
    isRepository: true,
    branch,
    ahead,
    behind,
    staged,
    modified,
    untracked,
    conflicted,
    clean: staged === 0 && modified === 0 && untracked === 0 && conflicted === 0,
    files,
  };
}

export async function collectGitSummary(
  root: string,
  status: GitStatusSnapshot,
): Promise<WorkspaceGitSummary> {
  const numstat = await runGit(root, [
    "diff",
    "--no-ext-diff",
    "--no-renames",
    "--numstat",
    "-z",
    "HEAD",
    "--",
  ]).catch(() => ({ stdout: "" }));
  const entries = new Map<string, WorkspaceGitFile>();
  for (const item of parseNumstat(numstat.stdout)) {
    const statusEntry = status.files.get(item.path);
    entries.set(item.path, {
      ...item,
      status: statusEntry?.status ?? "modified",
    });
  }
  for (const [relativePath, statusEntry] of status.files) {
    if (statusEntry.status !== "untracked" || entries.has(relativePath)) continue;
    try {
      const content = await readFile(path.join(root, relativePath), "utf8");
      entries.set(relativePath, {
        path: relativePath,
        status: "untracked",
        additions: countLines(content),
        deletions: 0,
      });
    } catch {
      entries.set(relativePath, {
        path: relativePath,
        status: "untracked",
        additions: null,
        deletions: null,
        binary: true,
      });
    }
  }
  for (const [filePath, statusEntry] of status.files) {
    if (!entries.has(filePath) && statusEntry.status !== "untracked") {
      entries.set(filePath, {
        path: filePath,
        status: statusEntry.status,
        additions: 0,
        deletions: 0,
        previousPath: statusEntry.previousPath,
      });
    }
  }
  for (const [filePath, statusEntry] of status.files) {
    if (!statusEntry.previousPath) continue;
    const current = entries.get(filePath);
    const previous = entries.get(statusEntry.previousPath);
    entries.set(filePath, {
      path: filePath,
      status: "renamed",
      additions: current?.additions ?? 0,
      deletions: previous?.deletions ?? current?.deletions ?? 0,
      previousPath: statusEntry.previousPath,
      binary: current?.binary || previous?.binary,
    });
    entries.delete(statusEntry.previousPath);
  }
  const allFiles = [...entries.values()];
  const files = allFiles.slice(0, MAX_GIT_FILES);
  return {
    branch: status.branch,
    upstream: null,
    ahead: status.ahead,
    behind: status.behind,
    insertions: allFiles.reduce((sum, file) => sum + (file.additions ?? 0), 0),
    deletions: allFiles.reduce((sum, file) => sum + (file.deletions ?? 0), 0),
    filesChanged: entries.size,
    files,
    truncated: entries.size > files.length,
  };
}

export async function readGitDiff(root: string, relativePath: string): Promise<WorkspaceGitDiff> {
  const safePath = safeRelativePath(root, relativePath);
  const status = await readGitStatus(root);
  const statusEntry = status.files.get(safePath);
  const numstat = await runGit(root, [
    "diff",
    "--no-ext-diff",
    "--no-renames",
    "--numstat",
    "-z",
    "HEAD",
    "--",
    safePath,
    ...(statusEntry?.previousPath ? [statusEntry.previousPath] : []),
  ]).catch(() => ({ stdout: "" }));
  const numstats = parseNumstat(numstat.stdout);
  const file = numstats.find((item) => item.path === safePath);
  const previousFile = statusEntry?.previousPath
    ? numstats.find((item) => item.path === statusEntry.previousPath)
    : undefined;
  const fileStatus = statusEntry?.status ?? "modified";
  const additions = file?.additions ?? (fileStatus === "untracked" ? 0 : null);
  const deletions = file?.deletions ?? previousFile?.deletions ?? 0;
  let content = "";
  let binary = file?.binary;
  if (fileStatus === "untracked") {
    try {
      const metadata = await stat(path.join(root, safePath));
      if (metadata.size <= MAX_GIT_DIFF_BYTES) {
        const source = await readFile(path.join(root, safePath), "utf8");
        content = source
          .split(/\r?\n/)
          .map((line) => `+${line}`)
          .join("\n");
        binary = false;
      } else {
        binary = true;
      }
    } catch {
      binary = true;
    }
  } else {
    const result = await runGit(
      root,
      ["diff", "--no-ext-diff", "--find-renames", "--unified=80", "HEAD", "--", safePath],
      MAX_GIT_DIFF_BYTES + 64 * 1024,
    ).catch(() => ({ stdout: "" }));
    content = result.stdout;
  }
  let truncated = false;
  if (Buffer.byteLength(content, "utf8") > MAX_GIT_DIFF_BYTES) {
    content = Buffer.from(content, "utf8").subarray(0, MAX_GIT_DIFF_BYTES).toString("utf8");
    truncated = true;
  }
  const originalContent = binary
    ? undefined
    : await readGitSnapshot(root, statusEntry?.previousPath ?? safePath);
  const modifiedContent = binary ? undefined : await readWorkingSnapshot(root, safePath);
  if (modifiedContent === null) truncated = true;
  return {
    path: safePath,
    previousPath: statusEntry?.previousPath,
    content,
    originalContent,
    modifiedContent: modifiedContent ?? undefined,
    additions,
    deletions,
    binary,
    truncated,
  };
}

export async function restoreGit(root: string, relativePath?: string): Promise<void> {
  if (relativePath?.trim()) {
    const safePath = safeRelativePath(root, relativePath);
    const status = await readGitStatus(root);
    const statusEntry = status.files.get(safePath);
    if (!statusEntry) return;
    if (statusEntry.status === "untracked") {
      await runGit(root, ["clean", "-fd", "--", safePath]);
      return;
    }
    if (statusEntry.status === "added") {
      await runGit(root, ["reset", "HEAD", "--", safePath]);
      await runGit(root, ["clean", "-fd", "--", safePath]);
      return;
    }
    const paths = [safePath];
    if (statusEntry.previousPath) {
      paths.unshift(safeRelativePath(root, statusEntry.previousPath));
    }
    await runGit(root, ["restore", "--source=HEAD", "--staged", "--worktree", "--", ...paths]);
    return;
  }
  await runGit(root, ["reset", "--hard", "HEAD"]);
  await runGit(root, ["clean", "-fd"]);
}

export async function commitGit(
  root: string,
  message?: string,
  push = false,
): Promise<WorkspaceGitCommitResult> {
  const commitMessage = message?.trim();
  if (!commitMessage) throw new Error("提交信息不能为空");
  await runGit(root, ["add", "-A"]);
  const status = await runGit(root, ["status", "--porcelain"]);
  if (!status.stdout.trim()) throw new Error("当前没有可提交的改动");
  await runGit(root, ["commit", "-m", commitMessage]);
  const hash = (await runGit(root, ["rev-parse", "HEAD"])).stdout.trim();
  if (push) await runGit(root, ["push"]);
  return { hash, message: commitMessage, pushed: push, generated: false };
}

export async function pushGit(root: string): Promise<WorkspaceGitCommitResult> {
  await runGit(root, ["push"]);
  const hash = (await runGit(root, ["rev-parse", "HEAD"])).stdout.trim();
  return { hash, message: "", pushed: true, generated: false };
}
